docs/design/1443-dynamic-pipelines.md

```diff
@@ -0,0 +1,252 @@
+  {"__address__" = "redis-two:9116", "instance" = "two"},
+]
+```
+
+The [Alloy type][alloy-types] of the list above is `list(map(string))`.
+However, you may want to pipe information from this list of targets to a component which doesn't work with a `list()` or a `map()`.
+For example, you may want to input the `"__address__"` string to a `prometheus.exporter.redis`,
+and you may want to use the `"instance"` string in a `discovery.relabel`.
+
+[alloy-types]: https://grafana.com/docs/alloy/latest/get-started/configuration-syntax/expressions/types_and_values/
+
+## Proposal 1: A foreach block
+
+A `foreach` block may start several sub-pipelines for a `collection` specified in its arguments.
+
+```alloy
+// All components in the sub-pipeline will be scoped under "foreach.default/1/...".
+// Here, "1" is sub-pipeline number 1.
+// This way component names won't clash with other sub-pipelines from the same foreach, 
+// and with the names of components outside of the foreach.
+foreach "default" {
+    
+  // "collection" is what the for loop will iterate over.
+  collection = discovery.file.default.targets
+
+  // Each item in the collection will be accessible via the "target" variable.
+  // E.g. `target["__address__"]`.
+  var = "target"
+
+  // A sub-pipeline consisting of components which process each target.
+  ...
+}
+```
+
+<details>
+  <summary>Example</summary>
+
+```alloy
+discovery.file "default" {
+  files = ["/Users/batman/Desktop/redis_addresses.yaml"]
+}
+
+// Every component defined in the "foreach" block will be instantiated for each item in the collection. 
+// The instantiated components will be scoped using the name of the foreach block and the index of the
+// item in the collection. For example: /foreach.redis/0/prometheus.exporter.redis.default
+foreach "redis" {
+  collection = discovery.file.default.targets
+  // Here, "target" is a variable whose value is the current item in the collection.
+  var = "target"
+
+  prometheus.exporter.redis "default" {
+    redis_addr = target["__address__"] // we can also do the necessary rewrites before this.
+  }
+
+  discovery.relabel "default" {
+    targets = prometheus.exporter.redis.default.targets
+    // Add a label which comes from the discovery component.
+    rule {
+      target_label = "filepath"
+      // __meta_filepath comes from discovery.file
+      replacement  = target["__meta_filepath"]
+    }
+  }
+
+  prometheus.scrape "default" {
+    targets = discovery.relabel.default.targets
+    forward_to = prometheus.remote_write.mimir.receiver
+  }
+}
+
+prometheus.remote_write "mimir" {
+  endpoint {
+    url = "https://prometheus-prod-05-gb-south-0.grafana.net/api/prom/push"
+    basic_auth {
+      username = ""
+      password = ""
+    }
+  }
+}
+```
+
+</details>
+
+Pros:
+* The `foreach` name is consistent with other programming languages.
+
+Cons:
+* It looks less like a component than a `declare.dynamic` block.
```

Comment:
I don't think it's a problem, you can still define a `declare` component (or import one) and use it inside the `foreach` syntax to make it smaller, add the parameters you need and reuse it for multiple foreach blocks.